R/seroprevalence_data.R

```diff
@@ -164,16 +192,39 @@ prepare_bin_data <- function(serodata) {
 #' foi <- rep(0.02, n_years)
 #' sim_probability <- get_sim_probability(sim_data = sim_data, foi=foi)
 #' @export
-get_sim_probability <- function(sim_data, foi) {
-  sim_data <- sim_data %>%
-    mutate(birth_year = .data$tsur - .data$age_mean_f)
+get_sim_probability <- function(
+  sim_data,
+  foi,
+  mu = 0,
+  model_type = "time-varying"
+  ) {
+  # Checks valid model specification
+stopifnot(
+  "model_type must be either 'time-varying' or 'age-varying'" =
+  model_type %in% c("time-varying", "age-varying")
+)
+
+  sim_data <- mutate(
+    sim_data,
+    birth_year = .data$tsur - .data$age_mean_f
+  )
   cohort_ages <- get_cohort_ages(sim_data)
-  exposure_ages <- rev(cohort_ages$age)
-  exposure_matrix <- get_exposure_matrix(sim_data) # nolint: object_usage_linter
-  probabilities <- 1 - exp(-drop(exposure_matrix %*% foi))
+  ages <- rev(cohort_ages$age)
+
+  if (model_type == "time-varying") {
+    exposure_matrix <- get_exposure_matrix(sim_data) # nolint: object_usage_linter
+    probabilities <-
+    (foi / (foi + mu)) * (1 - exp(-drop(exposure_matrix %*% (foi + mu))))
```

Comment:
I would put this into its own function (like we've done for age_varying) which gets its own unit tests; it also makes the code more humanly readable.

Reply:
Good point. I added `probability_exact_time_varying` to compute this probability. I'm computing the exposure matrix internally in the function to avoid the use of `get_exposure_matrix` and simplify the function usage.